### PR TITLE
feat: Add a ResettableLazy delegate.

### PR DESCRIPTION
**Pull request by bgrozev**

### CODE REVIEW

File: src/main/kotlin/org/jitsi/utils/Delegates.kt [hunk 1]
```diff
@@ -22,9 +22,29 @@ import kotlin.reflect.KProperty
...
+inline fun <T> observableWhenChanged(
+    initialValue: T,
+    crossinline onChange: () -> Unit
+): ReadWriteProperty<Any?, T> =
+    Delegates.observable(initialValue) { _, oldValue, newValue ->
+        if (oldValue != newValue) onChange()
+    }
+
+class ResettableLazy<T>(val initializer: () -> T) {
+    var lazyHolder = lazy { initializer() }
```
Comment (bbaldino) [inline src/main/kotlin/org/jitsi/utils/Delegates.kt:45]: Can this be private?

File: src/main/kotlin/org/jitsi/utils/Delegates.kt [hunk 2]
```diff
@@ -22,9 +22,29 @@ import kotlin.reflect.KProperty
...
+ */
+inline fun <T> observableWhenChanged(
+    initialValue: T,
+    crossinline onChange: () -> Unit
+): ReadWriteProperty<Any?, T> =
+    Delegates.observable(initialValue) { _, oldValue, newValue ->
+        if (oldValue != newValue) onChange()
+    }
+
+class ResettableLazy<T>(val initializer: () -> T) {
```
Comment (bbaldino) [inline src/main/kotlin/org/jitsi/utils/Delegates.kt:44]: `initializer` can also be private I think?

File: src/main/kotlin/org/jitsi/utils/Delegates.kt [hunk 3]
```diff
@@ -22,9 +22,29 @@ import kotlin.reflect.KProperty
...
+        if (oldValue != newValue) onChange(property, oldValue, newValue)
+    }
+/**
+ * A delegate which runs a callback (with no arguments) whenever the setter is called and it results in the value
+ * changing.
+ */
+inline fun <T> observableWhenChanged(
+    initialValue: T,
+    crossinline onChange: () -> Unit
+): ReadWriteProperty<Any?, T> =
```
Comment (bbaldino) [inline src/main/kotlin/org/jitsi/utils/Delegates.kt:39]: Could this re-use the other `observableWhenChanged`? Something like:

```
inline fun<T> observableWhenChanged(
    initialValue: T, 
    crossinline onChange: () -> Unit
) : ReadWriteProperty<Any?, T>) = observableWhenChanged(initialValue) { _, _, _ => onChange() }
```

**Comment by JonathanLennox** [comment]
I feel like this needs unit tests.

**Review by JonathanLennox — APPROVED** [review]

**Pull request merged**